x/wasm-storage/types/filters.go

```diff
@@ -0,0 +1,299 @@
+	switch f.numberType {
+	case 0x00: // Int32
+		return detectOutliersInteger[int32](dataList, f.maxSigma)
+	case 0x01: // Int64
+		return detectOutliersInteger[int64](dataList, f.maxSigma)
+	case 0x02: // Uint32
+		return detectOutliersInteger[uint32](dataList, f.maxSigma)
+	case 0x03: // Uint64
+		return detectOutliersInteger[uint64](dataList, f.maxSigma)
+	default:
+		return nil, ErrInvalidNumberType
+	}
+}
+
+func detectOutliersInteger[T constraints.Integer](dataList []string, maxSigma uint64) ([]int, error) {
+	var nums []T
+	var corruptCount int
+	var corruptInds []int
+	var z T
+	rt := reflect.TypeOf(z)
+	length := len(dataList)
+	for i, data := range dataList {
+		if data == "" {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+
+		bz, err := base64.StdEncoding.DecodeString(data)
+		if err != nil {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+		if len(bz) != int(rt.Size()) {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+
+		switch rt.Kind() {
+		case reflect.Int32:
+			var num int32
+			err = binary.Read(bytes.NewBuffer(bz), binary.BigEndian, &num)
+			if err != nil {
+				corruptCount++
+				corruptInds = append(corruptInds, i)
+				continue
+			}
+			nums = append(nums, T(num))
+
+		case reflect.Int64:
+			var num int64
+			err = binary.Read(bytes.NewBuffer(bz), binary.BigEndian, &num)
+			if err != nil {
+				corruptCount++
+				corruptInds = append(corruptInds, i)
+				continue
+			}
+			nums = append(nums, T(num))
+
+		case reflect.Uint32:
+			num := binary.BigEndian.Uint32(bz)
+			nums = append(nums, T(num))
+
+		case reflect.Uint64:
+			num := binary.BigEndian.Uint64(bz)
+			nums = append(nums, T(num))
+
+		default:
+			panic("invalid number type") // TODO should never end up here
+		}
+	}
+
+	// If more than 1/3 of the reveals are corrupted,
+	// we reach consensus that the reveals are unusable.
+	if corruptCount*3 > length {
+		return nil, ErrCorruptReveals
+	}
+
+	// Sort and find median.
+	numsSorted := make([]T, len(nums))
+	copy(numsSorted, nums)
+	slices.Sort(numsSorted)
+
+	var median T
+	var medianHalf bool // if true, median must be corrected by adding 0.5
+	if length%2 == 1 {
+		median = numsSorted[length/2]
+	} else {
+		mid := length / 2
+		median = (numsSorted[mid-1] + numsSorted[mid]) / 2
+		if (numsSorted[mid-1]%2 == 0 && numsSorted[mid]%2 == 1) ||
+			(numsSorted[mid-1]%2 == 1 && numsSorted[mid]%2 == 0) {
+			medianHalf = true
+		}
+	}
+
+	// Identify outliers and keep their count.
+	outliers := make([]int, len(dataList))
+	var nonOutlierCount int
+	i, j := 0, 0 // i tracks nums and j tracks corruptInds.
+	for i+j < len(dataList) {
+		if corruptCount > 0 && i+j == corruptInds[j] {
+			outliers[i+j] = 1
+			corruptCount--
+			j++
+			continue
+		}
+		if isOutlier(maxSigma, nums[i], median, medianHalf) {
+			outliers[i+j] = 1
+		} else {
+			nonOutlierCount++
+		}
+		i++
+	}
```

Comment:
I'm having a lil bit of a hard time understanding what is going on here. Can we maybe simplify this part (It doesn't matter if its less efficient)